defi05/python/dyosir.py

```diff
@@ -0,0 +1,148 @@
+        self.seuil = seuil
+        self.proba = proba
+        self.nouvel_etat = nouvel_etat
+        
+    def est_verifie(self, etats_voisines):
+        return etats_voisines.count(self.elt) >= self.seuil
+
+    
+class Monde:
+    def __init__(self, n, m, init = -1):
+        """
+        Pour initialiser aléatoirement : init = -1,
+        sinon pour initialiser avec un seul type de cellules,
+        init = état de la cellule.
+        """
+        self.n = n
+        self.m = m
+        if init == -1:
+            self.L = self.initialiser_aleatoirement()
+        else:
+            self.L = self.initialiser_etat(init)
+        self.iteration = 0
+        # regles[0] correspond aux événements possibles depuis l'état 0
+        # regles[1] aux événements possibles depuis l'état 1
+        # et ainsi de suite
+        self.regles = [
+            [Evenement(0, 0, 0.001, 1)],
+            [Evenement(3, 1, 0.001, 3), Evenement(4, 1, 0.002, 3), Evenement(5, 1, 0.001, 3), Evenement(0, 0, 0.005, 2)],
+            [Evenement(3, 1, 0.1, 3), Evenement(4, 1, 0.2, 3), Evenement(5, 1, 0.1, 3), Evenement(2, 5, 0.00005, 3)],
+            [Evenement(0, 0, 0.1, 4)],
+            [Evenement(0, 0, 0.1, 5)],
+            [Evenement(0, 0, 0.1, 0)],
+        ]
+        
+    def __str__(self):
+        affichage = ""
+        for l in self.L:
+            for c in l:
+                affichage += str(c)
+            affichage += "\n"
+        return affichage
+    
+    def effectifs(self):
+        eff = [0] * 6
+        for l in self.L:
+            for c in l:
+                eff[c.etat] += 1
+        return eff
+        
+    def initialiser_aleatoirement(self):
+        L = [[Cellule(randint(0, 5)) for _ in range(self.m)] for _ in range(self.n)]
+        return L
+    
+    def initialiser_etat(self, etat):
+        L = [[Cellule(etat) for _ in range(self.m)] for _ in range(self.n)]
+        return L
+        
+    def recuperer_etats_voisines(self, i, j):
+        """
+        Renvoie une liste de longueur variable contenant les états des cellules voisines.
+        Les cellules sur les bords ont moins de voisines, le monde ne boucle pas, donc
+        la liste renvoyée est de longueur variable.
+        """
+        # On génère {-1, 0, -1}^2
+        directions = {(x, y) for y in {-1, 0, 1} for x in {-1, 0, 1}}
+        # Puis on va enlever toutes les directions qui ne sont pas permises
+        # lorsqu'on est sur un bord
+        a_enlever = {(0, 0)}
+        if i == 0:
+            a_enlever.update({(-1, -1), (-1, 0), (-1, 1)})
+        if j == 0:
+            a_enlever.update({(-1, -1), (0, -1), (1, -1)})
+        if i == self.n - 1:
+            a_enlever.update({(1, -1), (1, 0), (1, 1)})
+        if j == self.m - 1:
+            a_enlever.update({(-1, 1), (0, 1), (1, 1)})
+        for e in a_enlever:
+            directions.remove(e)
+        etats = []
+        for d in directions:
+            etats.append(self.L[i + d[0]][j + d[1]].etat)
```

Comment:
c'est plus ici qu'il faut traiter le cas particulier de la bordure du monde, pas besoin de 100 conditions
(une autre solution est de fixer une rangée de cells a un etat neutre tout autour du monde et de ne pas itérer sur (0, n) mais sur (1, n-1) ce qui retire totalement cette préoccupation)